com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateAppStep.java

```diff
@@ -0,0 +1,419 @@
+import com.sap.cloud.lm.sl.cf.persistence.processors.DefaultFileDownloadProcessor;
+import com.sap.cloud.lm.sl.cf.persistence.services.FileContentProcessor;
+import com.sap.cloud.lm.sl.cf.persistence.services.FileStorageException;
+import com.sap.cloud.lm.sl.cf.process.Constants;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+import com.sap.cloud.lm.sl.cf.process.util.ApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ApplicationAttributeUpdater.UpdateState;
+import com.sap.cloud.lm.sl.cf.process.util.DiskQuotaApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ElementUpdater.UpdateBehavior;
+import com.sap.cloud.lm.sl.cf.process.util.EnvironmentApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.MemoryApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ServiceOperationUtil;
+import com.sap.cloud.lm.sl.cf.process.util.StagingApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.UrisApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.common.SLException;
+import com.sap.cloud.lm.sl.common.util.JsonUtil;
+import com.sap.cloud.lm.sl.common.util.MapUtil;
+import com.sap.cloud.lm.sl.mta.handlers.ArchiveHandler;
+import com.sap.cloud.lm.sl.mta.util.ValidatorUtil;
+
+@Component("createOrUpdateAppStep")
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class CreateOrUpdateAppStep extends SyncFlowableStep {
+
+    private SecureSerializationFacade secureSerializer = new SecureSerializationFacade();
+
+    @Autowired
+    protected ApplicationConfiguration configuration;
+
+    @Override
+    protected StepPhase executeStep(ExecutionWrapper execution) throws FileStorageException {
+        // Get the next cloud application from the context:
+        CloudApplicationExtended app = StepsUtil.getApp(execution.getContext());
+        try {
+            // Check if an application with this name already exists
+            CloudControllerClient client = execution.getControllerClient();
+            CloudApplication existingApp = client.getApplication(app.getName(), false);
+            StepsUtil.setExistingApp(execution.getContext(), existingApp);
+
+            boolean servicesPropertiesChanged = true;
+            boolean appPropertiesChanged = true;
+            boolean userPropertiesChanged = true;
+
+            printStepStartMessage(app, existingApp == null);
+
+            if (existingApp == null) {
+                createApplication(execution, client, app);
+            } else {
+                appPropertiesChanged = updateApplicationAttributes(app, existingApp, client);
+                reportApplicationUpdateStatus(app, appPropertiesChanged);
+            }
+
+            if (existingApp == null) {
+                bindServices(execution, client, app);
+            } else {
+                servicesPropertiesChanged = updateApplicationServices(app, existingApp, client, execution);

```

Comment:
Maybe add an aggrated general method e.g.'handleServices' that does the appropriate action; same for appEnv below